### ISSUE TITLE
Slots, slot groups and presets (| module signatures)

### ISSUE THREAD

**Issue by pshirshov** [7mind/izumi#292]
- _slot_ is an implementation alternative for a specific `DIKey`
- _preset_ is a map of `DIKey`s to preffered implementation
- _slot group_ is a set of several slots grouped by some criterias (both formal and informal). Slot groups may have intersections

1. Now we have a single place where we have access to all the plan conflicts (`PlanAnalyzerDefaultImpl`)
2. Each conlict effectively defines a slot
3. In case we have a _preset_ -  we may resolve conflicts with them
4. Right now we use tags with unsafe ahead-of-time binding pruning to resolve conflicts. We may make this logic more intelligent
5. We may try to find a better way to define/compute presets and to perform a check which would ensure that we have all the neccessary slot definitions 
6. In case we know all modules at compile time we may perform all the neccessary checks at compile time. Related: #51

**Comment by pshirshov** [comment]
An idea: 

```
trait Slot[+Id <: SlotId] {

}

trait SlotId



trait RepositorySlotOption extends SlotId

trait Production extends RepositorySlotOption

trait Repository[+Id <: RepositorySlotOption] extends Slot[Id]



class MyModule extends ModuleDef {
  protected def tagged[Id <: SlotId](body : => Unit) = ???

  tagged[Repository[Production]] {    
    make[Repo].from[RepoImpl]
  }

}
```

and then

```
./app -slot:repository=production
```

**Comment by pshirshov** [comment]
For now it's solved by #525 with independent axis (`-u env:test -u api:mock`) and unreachable conflict pruning.
